pkg/provider/naming.go

```diff
@@ -0,0 +1,44 @@
+package provider
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/pulumi/pulumi/pkg/tokens"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+var renameBlacklist = sets.NewString("v1/Namespace")
+
+// assignName generates a name for an object if it doesn't already have one. Uses DNS-1123-compliant
+// characters.
+func assignName(obj *unstructured.Unstructured, base tokens.QName) {
+	if !renameBlacklist.Has(obj.GroupVersionKind().String()) {
+		replaceNameIfUnnamed(obj, fmt.Sprintf("%s-%s", base, randString(8)))
+	} else {
+		replaceNameIfUnnamed(obj, string(base))
+	}
+}
+
+// replaceNameIfUnnamed will replace an object's name if it doesn't already have one.
+func replaceNameIfUnnamed(obj *unstructured.Unstructured, name string) {
```

Comment:
`setNameIfUnnamed` might be a better name--in a Pulumi context, `replace` has pretty specific connotations.